src/java/org/apache/cassandra/db/compaction/CompactionTask.java

```diff
@@ -188,26 +187,65 @@ protected void runMayThrow() throws Exception
         // it is not empty, it may compact down to nothing if all rows are deleted.
         assert transaction != null;
 
-        if (transaction.originals().isEmpty())
+        if (inputSSTables().isEmpty())
             return;
 
         if (DatabaseDescriptor.isSnapshotBeforeCompaction())
             realm.snapshotWithoutMemtable(System.currentTimeMillis() + "-compact-" + realm.getTableName());
 
         // The set of sstables given here may be later modified by buildCompactionCandidatesForAvailableDiskSpace() and
         // the compaction iterators in CompactionController and OverlapTracker will reflect the updated set of sstables.
-        try (CompactionController controller = getCompactionController(transaction.originals());
+        try (CompactionController controller = getCompactionController(inputSSTables());
              CompactionOperation operation = createCompactionOperation(controller, strategy))
         {
             operation.execute();
         }
     }
 
+    /**
+     * @return The token range that the operation should compact. This is usually null, but if we have a parallelizable
+     * multi-task operation (see {@link UnifiedCompactionStrategy#createAndAddTasks}), it will specify a subrange.
+     */
+    protected Range<Token> tokenRange()
+    {
+        return null;
+    }
```

Comment:
I see that the range is generated by the ShardManager, which indicates that maybe this token range is independent of the `InclusionMethod`.

Reply:
The range is indeed independent of the `InclusionMethod`. The range here is a range inside the input sstables, and however we select these we still split the output at predetermined positions.

In other words the `InclusionMethod` is a feature of the [compaction selection](https://github.com/datastax/cassandra/blob/STAR-1872/src/java/org/apache/cassandra/db/compaction/UnifiedCompactionStrategy.md#choosing-sstables-to-compact), while the range here is a feature of the [output sharding](https://github.com/datastax/cassandra/blob/STAR-1872/src/java/org/apache/cassandra/db/compaction/UnifiedCompactionStrategy.md#basic-sharding-scheme). The two are generally independent (and should be, to be able to correctly act in case of changes in sharding or upgrade from a legacy strategy), even though the latter should work in a way that makes the former efficient.

Reply:
Thanks for the helpful explanation